rlkit/samplers/in_place.py

```diff
@@ -13,32 +13,35 @@ class InPlacePathSampler(object):
     sampler.obtain_samples  # this has side-effects: env will change!
     ```
     """
-    def __init__(self, env, policy, max_samples, max_path_length):
+    def __init__(self, env, policy, max_path_length):
         self.env = env
         self.policy = policy
 
         self.max_path_length = max_path_length
-        self.max_samples = max_samples
-        assert (
-            max_samples >= max_path_length,
-            "Need max_samples >= max_path_length"
-        )
 
     def start_worker(self):
         pass
 
     def shutdown_worker(self):
         pass
 
-    def obtain_samples(self, deterministic=False, num_samples=None, resample=1):
+    def obtain_samples(self, deterministic=False, num_samples=None, num_trajs=None, resample=1):
+        """
+        Obtains samples in the environment until either we reach either num_samples transitions or
```

Comment:
do we ever not sample full trajectories in PEARL? seems like we could simplify quite a bit by only collecting full trajectories.

Reply:
It seems like the reason this is complicated is that during eval, the context is being updated online, and so only resampling `z` needs to be dealt with here. But in training, we want to recompute the posterior from sampled batches. Maybe the way to simplify is to have this function sample trajectories until it's time to resample `z`. Then whatever code is calling this can either sample a new `z` from the same posterior, or update the posterior as it wants to (batch or online) and then resample. The only problem here is that we lose the ability to update at the transition level...which we want for pomdps I suppose.

Reply:
even if we leave it like this, we should modify the `rollout` function to only accumulate context online during eval, it's a waste of computation during training.

Reply:
We now only sample full trajectories in PEARL, which should be what this function does.

Reply:
ok I can't figure out right now how to simplify this, so maybe we can just make it less confusing to read. Can we rename `num_samples` to `max_samples` and re-order the lines a bit (first deal with setting the max number of transitions, then deal with setting the max number of trajs in different code blocks) and add some explanatory comments?

Reply:
I've simplified it a bit to make it more readable.